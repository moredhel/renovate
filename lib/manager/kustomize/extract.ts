import { PackageFile, PackageDependency } from '../common';
import { extractBase, extractBases } from './common';
import { getDep } from '../dockerfile/extract';
import { logger } from '../../logger';

export function extractPackageFile(content: string): PackageFile | null {
  logger.trace('kustomize.extractPackageFile()');
  let deps: PackageDependency[] = [];
  let lineNumber = 0;

  let pkg = extractBases(content);
  if (!pkg) {
    return null;
  }

  // grab the remote bases
  for (const base of pkg.bases) {
    const dep = extractBase(base);
    if (dep) {
      deps.push(dep);
    }
  }

  if (!deps.length) {
    return null;
  }
  return { deps };
}
