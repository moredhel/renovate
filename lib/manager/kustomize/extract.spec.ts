import { readFileSync } from 'fs';
import { extractPackageFile } from './extract';

const kustomizeGitSSHBase = readFileSync(
  'lib/manager/kustomize/__fixtures__/gitSshBase.yaml',
  'utf8'
);

const kustomizeGitSSHSubdir = readFileSync(
  'lib/manager/kustomize/__fixtures__/gitSubdir.yaml',
  'utf8'
);

const kustomizeHTTP = readFileSync(
  'lib/manager/kustomize/__fixtures__/kustomizeHttp.yaml',
  'utf8'
);

const kustomizeWithLocal = readFileSync(
  'lib/manager/kustomize/__fixtures__/kustomizeWithLocal.yaml',
  'utf8'
);

const nonKustomize = readFileSync(
  'lib/manager/kustomize/__fixtures__/service.yaml',
  'utf8'
);

describe('lib/manager/kustomize/extract', () => {
  describe('extractPackageFile()', () => {
    it('returns null for non kustomize kubernetes files', () => {
      expect(extractPackageFile(nonKustomize)).toBeNull();
    });
    it('extracts multiple image lines', () => {
      const res = extractPackageFile(kustomizeHTTP);
      expect(res.deps).toMatchSnapshot();
      expect(res.deps).toHaveLength(2);
    });
    it('extracts ssh dependency', () => {
      const res = extractPackageFile(kustomizeGitSSHBase);
      expect(res.deps).toMatchSnapshot();
      expect(res.deps).toHaveLength(1);
    });
    it('extracts ssh dependency with a subdir', () => {
      const res = extractPackageFile(kustomizeGitSSHSubdir);
      expect(res.deps).toMatchSnapshot();
      expect(res.deps).toHaveLength(1);
    });
    it('extracts http dependency', () => {
      const res = extractPackageFile(kustomizeHTTP);
      expect(res.deps).toMatchSnapshot();
      expect(res.deps).toHaveLength(2);
      expect(res.deps[1].source).toEqual(
        'https://github.com/moredhel/remote-kustomize.git'
      );
      expect(res.deps[1].currentValue).toEqual('v0.0.1');
    });

    it('ignores non-Kubernetes empty files', () => {
      expect(extractPackageFile('')).toBeNull();
    });
  });
});
