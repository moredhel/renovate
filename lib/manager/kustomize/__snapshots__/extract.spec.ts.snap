// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`lib/manager/kustomize/extract extractPackageFile() extracts http dependency 1`] = `
Array [
  Object {
    "currentValue": "v0.0.1",
    "datasource": "git-tags",
    "depName": "github.com/user/repo//deploy",
    "depType": "git-tags",
    "lookupName": "github.com/user/repo",
    "source": "github.com/user/repo",
  },
]
`;

exports[`lib/manager/kustomize/extract extractPackageFile() extracts multiple image lines 1`] = `
Array [
  Object {
    "currentValue": "v0.0.1",
    "datasource": "git-tags",
    "depName": "https://moredhel/remote-kustomize.git",
    "depType": "git-tags",
    "lookupName": "https://moredhel/remote-kustomize.git",
    "source": "https://moredhel/remote-kustomize.git",
  },
  Object {
    "currentValue": "v0.0.1",
    "datasource": "git-tags",
    "depName": "https://moredhel/remote-kustomize.git//deploy",
    "depType": "git-tags",
    "lookupName": "https://moredhel/remote-kustomize.git//deploy",
    "source": "https://moredhel/remote-kustomize.git//deploy",
  },
]
`;

exports[`lib/manager/kustomize/extract extractPackageFile() extracts ssh dependency 1`] = `
Array [
  Object {
    "currentValue": "v0.0.1",
    "datasource": "git-tags",
    "depName": "git@github.com:moredhel/remote-kustomize.git",
    "depType": "git-tags",
    "lookupName": "git@github.com:moredhel/remote-kustomize.git",
    "source": "git@github.com:moredhel/remote-kustomize.git",
  },
]
`;

exports[`lib/manager/kustomize/extract extractPackageFile() extracts ssh dependency with a subdir 1`] = `
Array [
  Object {
    "currentValue": "v2.0.0",
    "datasource": "git-tags",
    "depName": "git@github.com:kubernetes-sigs/kustomize.git//examples/helloWorld",
    "depType": "git-tags",
    "lookupName": "git@github.com:kubernetes-sigs/kustomize.git",
    "source": "git@github.com:kubernetes-sigs/kustomize.git",
  },
]
`;

exports[`lib/manager/kustomize/extract extractPackageFile() should extract out image versions 1`] = `
Array [
  Object {
    "currentValue": "v0.1.0",
    "datasource": "docker",
    "depName": "node",
    "depType": "docker",
    "lookupName": "node",
    "source": "node",
  },
  Object {
    "currentValue": "v0.0.1",
    "datasource": "docker",
    "depName": "group/instance",
    "depType": "docker",
    "lookupName": "group/instance",
    "source": "group/instance",
  },
  Object {
    "currentValue": "v0.0.2",
    "datasource": "docker",
    "depName": "quay.io/test/repo",
    "depType": "docker",
    "lookupName": "quay.io/test/repo",
    "source": "quay.io/test/repo",
  },
  Object {
    "currentValue": "v0.0.3",
    "datasource": "docker",
    "depName": "gitlab.com/org/suborg/image",
    "depType": "docker",
    "lookupName": "gitlab.com/org/suborg/image",
    "source": "gitlab.com/org/suborg/image",
  },
]
`;
